<a href="https://colab.research.google.com/github/SrijaKandese21uari200/SPAM_DETECTION_NLP_TEAM-30/blob/main/Spam_Detection_Team-30.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**NECESSARY LIBRARIES**

In [29]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import requests
from io import BytesIO
import zipfile


**DATASET( WE HAVE SELECTED THIS WITH THE HELP OF CHAT GPT)**

In [30]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00228/smsspamcollection.zip"
r = requests.get(url)
with zipfile.ZipFile(BytesIO(r.content), 'r') as z:
    with z.open('SMSSpamCollection') as f:
        df = pd.read_csv(f, sep='\t', names=['label', 'message'])



The dataset we selected has collection of emails which are divided into spam and ham(non-spam).It is one of the most reliable data sets, which covers maximum all the cases. To proceed further we labeled **'ham' to 0 and 'spam' to 1**

In [31]:
df['label'] = df['label'].map({'ham': 0, 'spam': 1})


We split the dataset into **training and testing sets**

In [32]:
X_train, X_test, y_train, y_test = train_test_split(df['message'], df['label'], test_size=0.2, random_state=42)


We used **Count Vectorizer** to convert collections of text documents (both train and test datasets) into matrices of token counts. Here we used **fit_transform ** to load the training data.Further we used **transform** to  transform testing data in the same way as the training data.

In [33]:
Vectorizer = CountVectorizer()
X_train = Vectorizer.fit_transform(X_train)
X_test = Vectorizer.transform(X_test)



**WE USED NAIVE BAYES CLASSIFIER**

In [34]:
used_classifier = MultinomialNB()
used_classifier.fit(X_train, y_train)


MultinomialNB()

In [35]:
y_pred = used_classifier.predict(X_test)

**PERFORMANCE OF MODEL CONSIDERED**

In [36]:
Accuracy = accuracy_score(y_test, y_pred)
confusion_matrix = confusion_matrix(y_test.tolist(), y_pred.tolist())
classification_report = classification_report(y_test, y_pred)

print(f"Accuracy: {Accuracy}")
print(f"Confusion Matrix:\n{confusion_matrix}")
print(f"Classification Report:\n{classification_report}")

#Output

Accuracy: 0.9919282511210762
Confusion Matrix:
[[966   0]
 [  9 140]]
Classification Report:
              precision    recall  f1-score   support

           0       0.99      1.00      1.00       966
           1       1.00      0.94      0.97       149

    accuracy                           0.99      1115
   macro avg       1.00      0.97      0.98      1115
weighted avg       0.99      0.99      0.99      1115



We calculated the performance of our model before taking a test case. The **accuracy is 99.19%**. From the confusion matrix if **does not have false positives**, i.e it does not classify any non-spam message as spam message. Our main goal is to to ensure that no non-spam message is going to spam section, which is true in our model. We have also calculated the classification report.

**EXAMPLES**

In [37]:
# Example 1
new_email = ["Congratulations! You've won a lottery."]
new_email_transformed = Vectorizer.transform(new_email)
prediction = used_classifier.predict(new_email_transformed)

print("Predictions for new emails:",prediction)

Predictions for new emails: [1]


The above example is of SPAM

In [38]:
# Example 2
new_email = ["Meeting scheduled for tomorrow."]
new_email_transformed = Vectorizer.transform(new_email)
prediction = used_classifier.predict(new_email_transformed)

print("Predictions for new emails:",prediction)

Predictions for new emails: [0]


The above example is of NON_SPAM

In [39]:
# Example 3
new_email = ["50% discount on electronic devices!!!!!"]
new_email_transformed = Vectorizer.transform(new_email)
prediction = used_classifier.predict(new_email_transformed)

print("Predictions for new emails:",prediction)

Predictions for new emails: [1]


The above example is of SPAM

In [40]:
# Example 4
new_email = ["Parent-Teachers meet tomorrow,must attend!"]
new_email_transformed = Vectorizer.transform(new_email)
prediction = used_classifier.predict(new_email_transformed)

print("Predictions for new emails:",prediction)

Predictions for new emails: [0]


The above example is of NON_SPAM